File: src/indicators.py
"""Technical indicator calculations."""

import pandas as pd
import logging
from typing import Dict, Optional, Tuple

try:
    import pandas_ta as ta
except Exception:  # pragma: no cover - fallback keeps tests usable without pandas-ta
    ta = None

logger = logging.getLogger(__name__)

class TechnicalIndicators:
    """Calculate and manage technical indicators"""
    
    @staticmethod
    def calculate_rsi(
        df: pd.DataFrame, 
        period: int = 14, 
        column: str = 'close'
    ) -> pd.Series:
        """
        Calculate Relative Strength Index (RSI)
        
        Args:
            df: DataFrame with price data
            period: RSI period (default 14)
            column: Column to calculate RSI on
            
        Returns:
            Series with RSI values
        """
        if df.empty or column not in df.columns:
            return pd.Series()
        
        if ta is not None:
            rsi = ta.rsi(df[column], length=period)
            if rsi is not None:
                return rsi

        delta = df[column].diff()
        gain = delta.clip(lower=0).ewm(alpha=1 / period, adjust=False).mean()
        loss = (-delta.clip(upper=0)).ewm(alpha=1 / period, adjust=False).mean()
        rs = gain / loss.replace(0, pd.NA)
        return 100 - (100 / (1 + rs))
    
    @staticmethod
    def calculate_macd(
        df: pd.DataFrame,
        fast: int = 12,
        slow: int = 26,
        signal: int = 9,
        column: str = 'close'
    ) -> Tuple[pd.Series, pd.Series, pd.Series]:
        """
        Calculate MACD (Moving Average Convergence Divergence)
        
        Args:
            df: DataFrame with price data
            fast: Fast EMA period
            slow: Slow EMA period
            signal: Signal line period
            column: Column to calculate on
            
        Returns:
            Tuple of (MACD, Signal, Histogram)
        """
        if df.empty or column not in df.columns:
            return pd.Series(), pd.Series(), pd.Series()
        
        if ta is not None:
            macd = ta.macd(df[column], fast=fast, slow=slow, signal=signal)
            if macd is not None and not macd.empty:
                macd_line = macd.iloc[:, 0] if macd.shape[1] > 0 else pd.Series(index=df.index)
                histogram = macd.iloc[:, 1] if macd.shape[1] > 1 else pd.Series(index=df.index)
                signal_line = macd.iloc[:, 2] if macd.shape[1] > 2 else pd.Series(index=df.index)
                return macd_line, signal_line, histogram

        ema_fast = df[column].ewm(span=fast, adjust=False).mean()
        ema_slow = df[column].ewm(span=slow, adjust=False).mean()
        macd_line = ema_fast - ema_slow
        signal_line = macd_line.ewm(span=signal, adjust=False).mean()
        histogram = macd_line - signal_line
        return macd_line, signal_line, histogram
    
    @staticmethod
    def calculate_bollinger_bands(
        df: pd.DataFrame,
        period: int = 20,
        std_dev: float = 2.0,
        column: str = 'close'
    ) -> Tuple[pd.Series, pd.Series, pd.Series]:
        """
        Calculate Bollinger Bands
        
        Args:
            df: DataFrame with price data
            period: Moving average period
            std_dev: Standard deviation multiplier
            column: Column to calculate on
            
        Returns:
            Tuple of (Upper Band, Middle Band, Lower Band)
        """
        if df.empty or column not in df.columns:
            return pd.Series(), pd.Series(), pd.Series()
        
        if ta is not None:
            bb = ta.bbands(df[column], length=period, std=std_dev)
            if bb is not None and not bb.empty:
                upper = bb.iloc[:, 2] if bb.shape[1] > 2 else pd.Series(index=df.index)
                middle = bb.iloc[:, 1] if bb.shape[1] > 1 else pd.Series(index=df.index)
                lower = bb.iloc[:, 0] if bb.shape[1] > 0 else pd.Series(index=df.index)
                return upper, middle, lower

        middle = df[column].rolling(period).mean()
        std = df[column].rolling(period).std()
        upper = middle + std_dev * std
        lower = middle - std_dev * std
        return upper, middle, lower
    
    @staticmethod
    def calculate_volume_weighted_average_price(
        df: pd.DataFrame
    ) -> pd.Series:
        """
        Calculate VWAP (Volume Weighted Average Price)
        
        Args:
            df: DataFrame with price and volume data
            
        Returns:
            Series with VWAP values
        """
        if df.empty or 'close' not in df.columns or 'volume' not in df.columns:
            return pd.Series()
        
        high = df['high'] if 'high' in df.columns else df['close']
        low = df['low'] if 'low' in df.columns else df['close']
        typical_price = (high + low + df['close']) / 3

        if ta is not None:
            try:
                vwap = ta.vwap(high=high, low=low, close=df['close'], volume=df['volume'])
                if vwap is not None:
                    return vwap
            except Exception as e:
                logger.debug(f"pandas-ta VWAP failed, using fallback: {e}")

        volume = df['volume'].replace(0, pd.NA)
        return (typical_price * volume).cumsum() / volume.cumsum()

    @staticmethod
    def calculate_atr(df: pd.DataFrame, period: int = 14) -> pd.Series:
        """Calculate Average True Range."""
        required = {'high', 'low', 'close'}
        if df.empty or not required.issubset(df.columns):
            return pd.Series()

        if ta is not None:
            atr = ta.atr(df['high'], df['low'], df['close'], length=period)
            if atr is not None:
                return atr

        previous_close = df['close'].shift(1)
        tr = pd.concat(
            [
                df['high'] - df['low'],
                (df['high'] - previous_close).abs(),
                (df['low'] - previous_close).abs(),
            ],
            axis=1,
        ).max(axis=1)
        return tr.rolling(period).mean()
    
    @staticmethod
    def calculate_kdj(
        df: pd.DataFrame,
        period: int = 9,
        k_smooth: int = 3,
        d_smooth: int = 3,
    ) -> Tuple[pd.Series, pd.Series, pd.Series]:
        """
        Calculate KDJ (Stochastic Oscillator)
        
        Args:
            df: DataFrame with OHLCV data
            period: Lookback period (default 9)
            k_smooth: K line smoothing period (default 3)
            d_smooth: D line smoothing period (default 3)
            
        Returns:
            Tuple of (K, D, J) series
        """
        try:
            if df.empty or 'high' not in df.columns or 'low' not in df.columns or 'close' not in df.columns:
                return pd.Series(), pd.Series(), pd.Series()
            
            if ta is not None:
                stoch = ta.stoch(
                    high=df['high'],
                    low=df['low'],
                    close=df['close'],
                    k=period,
                    d=d_smooth,
                    smooth_k=k_smooth,
                )
                if stoch is not None and not stoch.empty:
                    k_line = stoch.iloc[:, 0] if stoch.shape[1] > 0 else pd.Series(index=df.index)
                    d_line = stoch.iloc[:, 1] if stoch.shape[1] > 1 else pd.Series(index=df.index)
                    j_line = 3 * k_line - 2 * d_line
                    return k_line, d_line, j_line

            lowest_low = df['low'].rolling(period).min()
            highest_high = df['high'].rolling(period).max()
            rsv = (df['close'] - lowest_low) / (highest_high - lowest_low).replace(0, pd.NA) * 100
            k_line = rsv.rolling(k_smooth).mean()
            d_line = k_line.rolling(d_smooth).mean()
            j_line = 3 * k_line - 2 * d_line
            return k_line, d_line, j_line
            
        except Exception as e:
            logger.error(f"Error calculating KDJ: {str(e)}")
            return pd.Series(), pd.Series(), pd.Series()
    
    @staticmethod
    def add_all_indicators(
        df: pd.DataFrame,
        rsi_period: int = 14,
        macd_fast: int = 12,
        macd_slow: int = 26,
        macd_signal: int = 9,
        bb_period: int = 20,
        bb_std: float = 2.0,
        kdj_period: int = 9,
        atr_period: int = 14,
    ) -> pd.DataFrame:
        """
        Add all technical indicators to DataFrame
        
        Args:
            df: DataFrame with OHLCV data
            rsi_period: RSI period
            macd_fast: MACD fast period
            macd_slow: MACD slow period
            macd_signal: MACD signal period
            bb_period: Bollinger Band period
            bb_std: Bollinger Band standard deviation
            
        Returns:
            DataFrame with indicators added
        """
        if df.empty:
            return df
        
        df_copy = df.copy()
        
        # Normalize column names
        df_copy.columns = [col.lower() for col in df_copy.columns]
        
        # Calculate RSI
        df_copy['rsi'] = TechnicalIndicators.calculate_rsi(
            df_copy, period=rsi_period
        )
        
        # Calculate MACD
        macd, signal, histogram = TechnicalIndicators.calculate_macd(
            df_copy, fast=macd_fast, slow=macd_slow, signal=macd_signal
        )
        df_copy['macd'] = macd
        df_copy['macd_signal'] = signal
        df_copy['macd_histogram'] = histogram
        
        # Calculate Bollinger Bands
        upper, middle, lower = TechnicalIndicators.calculate_bollinger_bands(
            df_copy, period=bb_period, std_dev=bb_std
        )
        df_copy['bb_upper'] = upper
        df_copy['bb_middle'] = middle
        df_copy['bb_lower'] = lower
        
        # Calculate KDJ
        k_line, d_line, j_line = TechnicalIndicators.calculate_kdj(
            df_copy, period=kdj_period
        )
        df_copy['kdj_k'] = k_line
        df_copy['kdj_d'] = d_line
        df_copy['kdj_j'] = j_line
        
        # Calculate VWAP
        df_copy['vwap'] = TechnicalIndicators.calculate_volume_weighted_average_price(
            df_copy
        )

        # Calculate ATR
        df_copy['atr'] = TechnicalIndicators.calculate_atr(df_copy, period=atr_period)
        
        return df_copy

    @staticmethod
    def calculate_all_indicators(
        df: pd.DataFrame,
        config: Optional[Dict] = None,
        **kwargs,
    ) -> pd.DataFrame:
        """Compatibility wrapper used by main/tests."""
        params = {}
        if config:
            params.update(config)
        params.update(kwargs)

        return TechnicalIndicators.add_all_indicators(
            df,
            rsi_period=params.get('rsi_period', 14),
            macd_fast=params.get('macd_fast', 12),
            macd_slow=params.get('macd_slow', 26),
            macd_signal=params.get('macd_signal', 9),
            bb_period=params.get('bb_period', 20),
            bb_std=params.get('bb_std', 2.0),
            kdj_period=params.get('kdj_period', 9),
            atr_period=params.get('atr_period', 14),
        )

File: src/indicators.cpp
#include "trading/indicators.hpp"

#include <algorithm>
#include <cmath>
#include <numeric>

namespace trading {

namespace {

constexpr double kEpsilon = 1e-9;

double ema_next(double value, double previous, int period)
{
    const double alpha = 2.0 / (static_cast<double>(period) + 1.0);
    return alpha * value + (1.0 - alpha) * previous;
}

double stddev_window(const std::vector<double>& values, std::size_t end_index, std::size_t period)
{
    if (end_index + 1 < period || period == 0) {
        return 0.0;
    }

    const std::size_t begin = end_index + 1 - period;
    const double mean = simple_moving_average(values, end_index, period);
    double sum = 0.0;
    for (std::size_t i = begin; i <= end_index; ++i) {
        const double diff = values[i] - mean;
        sum += diff * diff;
    }
    return std::sqrt(sum / static_cast<double>(period));
}

} // namespace

double simple_moving_average(const std::vector<double>& values,
                             std::size_t end_index,
                             std::size_t period)
{
    if (values.empty() || period == 0) {
        return 0.0;
    }
    const std::size_t count = std::min(period, end_index + 1);
    const std::size_t begin = end_index + 1 - count;
    const double sum = std::accumulate(values.begin() + static_cast<std::ptrdiff_t>(begin),
                                       values.begin() + static_cast<std::ptrdiff_t>(end_index + 1),
                                       0.0);
    return sum / static_cast<double>(count);
}

std::vector<IndicatorSnapshot> compute_indicators(const std::vector<Bar>& bars,
                                                  const StrategyConfig& config)
{
    std::vector<IndicatorSnapshot> snapshots(bars.size());
    if (bars.empty()) {
        return snapshots;
    }

    std::vector<double> closes;
    closes.reserve(bars.size());
    for (const auto& bar : bars) {
        closes.push_back(bar.close);
    }

    double fast_ema = closes.front();
    double slow_ema = closes.front();
    double signal_ema = 0.0;
    double avg_gain = 0.0;
    double avg_loss = 0.0;
    double atr = 0.0;
    double cumulative_pv = 0.0;
    double cumulative_volume = 0.0;
    double k_prev = 50.0;
    double d_prev = 50.0;

    for (std::size_t i = 0; i < bars.size(); ++i) {
        auto& out = snapshots[i];
        const Bar& bar = bars[i];
        const double typical_price = (bar.high + bar.low + bar.close) / 3.0;
        cumulative_pv += typical_price * bar.volume;
        cumulative_volume += bar.volume;
        out.vwap = cumulative_volume > kEpsilon ? cumulative_pv / cumulative_volume : bar.close;

        fast_ema = i == 0 ? bar.close : ema_next(bar.close, fast_ema, config.macd_fast);
        slow_ema = i == 0 ? bar.close : ema_next(bar.close, slow_ema, config.macd_slow);
        out.macd = fast_ema - slow_ema;
        signal_ema = i == 0 ? out.macd : ema_next(out.macd, signal_ema, config.macd_signal);
        out.macd_signal = signal_ema;
        out.macd_histogram = out.macd - out.macd_signal;

        out.bollinger_middle = simple_moving_average(closes, i, static_cast<std::size_t>(config.bollinger_period));
        const double deviation = stddev_window(closes, i, static_cast<std::size_t>(config.bollinger_period));
        out.bollinger_upper = out.bollinger_middle + config.bollinger_stddev * deviation;
        out.bollinger_lower = out.bollinger_middle - config.bollinger_stddev * deviation;

        if (i > 0) {
            const double change = bar.close - bars[i - 1].close;
            const double gain = std::max(change, 0.0);
            const double loss = std::max(-change, 0.0);
            if (i <= static_cast<std::size_t>(config.rsi_period)) {
                avg_gain += gain;
                avg_loss += loss;
                if (i == static_cast<std::size_t>(config.rsi_period)) {
                    avg_gain /= static_cast<double>(config.rsi_period);
                    avg_loss /= static_cast<double>(config.rsi_period);
                }
            } else {
                avg_gain = (avg_gain * (config.rsi_period - 1) + gain) / static_cast<double>(config.rsi_period);
                avg_loss = (avg_loss * (config.rsi_period - 1) + loss) / static_cast<double>(config.rsi_period);
            }

            if (i >= static_cast<std::size_t>(config.rsi_period)) {
                if (avg_loss < kEpsilon) {
                    out.rsi = 100.0;
                } else {
                    const double rs = avg_gain / avg_loss;
                    out.rsi = 100.0 - (100.0 / (1.0 + rs));
                }
            }
        }

        if (i == 0) {
            atr = bar.high - bar.low;
        } else {
            const double tr = std::max({bar.high - bar.low,
                                        std::abs(bar.high - bars[i - 1].close),
                                        std::abs(bar.low - bars[i - 1].close)});
            if (i < static_cast<std::size_t>(config.atr_period)) {
                atr = ((atr * static_cast<double>(i)) + tr) / static_cast<double>(i + 1);
            } else {
                atr = ((atr * static_cast<double>(config.atr_period - 1)) + tr) /
                      static_cast<double>(config.atr_period);
            }
        }
        out.atr = atr;

        const std::size_t kdj_period = static_cast<std::size_t>(config.kdj_period);
        const std::size_t begin = i + 1 > kdj_period ? i + 1 - kdj_period : 0;
        double highest = bars[begin].high;
        double lowest = bars[begin].low;
        for (std::size_t j = begin; j <= i; ++j) {
            highest = std::max(highest, bars[j].high);
            lowest = std::min(lowest, bars[j].low);
        }
        const double rsv = (highest - lowest) > kEpsilon
            ? (bar.close - lowest) / (highest - lowest) * 100.0
            : 50.0;
        out.k = (2.0 * k_prev + rsv) / 3.0;
        out.d = (2.0 * d_prev + out.k) / 3.0;
        out.j = 3.0 * out.k - 2.0 * out.d;
        k_prev = out.k;
        d_prev = out.d;
    }

    return snapshots;
}

} // namespace trading
